scipy/optimize/tnc/_tnc_module.pyx

```diff
@@ -0,0 +1,204 @@
+# cython: language_level=3, boundscheck=False
+
+from libc.string cimport memcpy
+import numpy as np
+cimport numpy as np
+cimport cython
+
+
+np.import_array()
+
+ctypedef np.float64_t float64_t
+
+
+cdef extern from "tnc.h":
+    ctypedef void tnc_callback(double[], void*) except *
+    ctypedef int tnc_function(double[], double*, double[], void*) except 1
+
+    cdef enum tnc_rc:
+        TNC_MINRC
+        TNC_ENOMEM
+        TNC_EINVAL
+        TNC_INFEASIBLE
+        TNC_LOCALMINIMUM
+        TNC_FCONVERGED
+        TNC_XCONVERGED
+        TNC_MAXFUN
+        TNC_LSFAIL
+        TNC_CONSTANT
+        TNC_NOPROGRESS
+        TNC_USERABORT
+
+    cdef int tnc(int n, double x[], double *f, double g[],
+            tnc_function *function, void *state,
+            double low[], double up[], double scale[], double offset[],
+            int messages, int maxCGit, int maxnfeval, double eta, double stepmx,
+            double accuracy, double fmin, double ftol, double xtol, double pgtol,
+            double rescale, int *nfeval, int *niter, tnc_callback *tnc_callback) except *
+
+
+cdef struct s_pytnc_state:
+  void *py_function
+  void *py_callback
+  int n
+  int failed
+ctypedef s_pytnc_state pytnc_state
+
+
+cdef int function(double x[], double *f, double g[], void *state) except 1:
+    # return 0 if no error, 1 if error
+    # However, if the user function raises Exceptions Cython should deal
+    # with it.
+
+    cdef:
+        pytnc_state *py_state
+        int n
+        double *x_data
+        double *g_data
+
+    py_state = <pytnc_state *>state
+    n = py_state.n
+
+    # ensures we're working on a copy of the data, just in case user function
+    # mutates it
+    xcopy = np.empty(n, dtype=np.float64)
+    x_data = <float64_t *>np.PyArray_DATA(xcopy)
+    memcpy(x_data, x, sizeof(double) * n)
+
+    fx, gx = (<object>py_state.py_function)(xcopy)
+
+    if not np.isscalar(fx):
+        try:
+            fx = np.asarray(fx).item()
+        except (TypeError, ValueError) as e:
+            raise ValueError(
+                "The user-provided objective function "
+                "must return a scalar value."
+            ) from e
+
+    f[0] = <double> fx
+
+    gx = np.asfarray(gx)
+    if gx.size != n:
+        raise ValueError("tnc: gradient must have shape (len(x0),)")
+
+    if not gx.flags['C_CONTIGUOUS']:
+        gx = np.ascontiguousarray(gx, dtype=np.float64)
+
+    g_data = <float64_t *>np.PyArray_DATA(gx)
+    memcpy(g, g_data, n * sizeof(double))
+
+    return 0
+
+
+cdef void callback_function(double x[], void *state) except *:
```

Comment:
There's an issue here that I didn't pick up on earlier (and probably should have done - sorry!).

Looking at the `tnc` code - there's no way for this function to indicate that it's failed (and therefore it can't stop the overall operation). This will potentially cause problems later when some other bit of Python code runs and finds that an exception is already set.

I think the solution is to add (at the start of `function`):

```
if py_state.failed:
    return 1
```

i.e. to have `function` check to make sure that `callback_function` hasn't failed on the previous call.

It's a little dodgy because it isn't handling the exceptions how quite Cython expects, but I think it should work providing it happens before anything Python-like is executed. I'm slightly surprised that the exception test currently passes.

Reply:
This also breaks with the current implementation in scipy:

```
>>> from scipy import optimize
>>> def callback(*a):
...   raise RuntimeError("xxx")
... 
>>> optimize.minimize(lambda x: optimize.rosen(x), [0, 1], method="TNC", callback=callback)
Traceback (most recent call last):
  File "<stdin>", line 2, in callback
RuntimeError: xxx

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.8/site-packages/scipy/optimize/_minimize.py", line 622, in minimize
    return _minimize_tnc(fun, x0, args, jac, bounds, callback=callback,
  File "/usr/lib64/python3.8/site-packages/scipy/optimize/tnc.py", line 414, in _minimize_tnc
    rc, nf, nit, x = moduleTNC.minimize(func_and_grad, x0, low, up, scale,
  File "/usr/lib64/python3.8/site-packages/scipy/optimize/_differentiable_functions.py", line 258, in fun_and_grad
    if not np.array_equal(x, self.x):
  File "<__array_function__ internals>", line 4, in array_equal
SystemError: PyEval_EvalFrameEx returned a result with an error set
```

The other option would just be to have `callback` swallow the exception (maybe printing it first). To be honest, that might be cleaner and simpler

Reply:
When I run your example (with the PR in its current state) I get:

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
/var/folders/m8/gwtcncws12jf60xw5n6knwnw0000gn/T/ipykernel_19744/2729457978.py in <module>
      3   raise RuntimeError("xxx")
      4 
----> 5 optimize.minimize(lambda x: optimize.rosen(x), [0, 1], method="TNC", callback=callback)

~/Documents/Andy/programming/scipy/scipy/optimize/_minimize.py in minimize(fun, x0, args, method, jac, hess, hessp, bounds, constraints, tol, callback, options)
    665                                callback=callback, **options)
    666     elif meth == 'tnc':
--> 667         res = _minimize_tnc(fun, x0, args, jac, bounds, callback=callback,
    668                             **options)
    669     elif meth == 'cobyla':

~/Documents/Andy/programming/scipy/scipy/optimize/_tnc.py in _minimize_tnc(fun, x0, args, jac, bounds, eps, scale, offset, mesg_num, maxCGit, maxiter, eta, stepmx, accuracy, minfev, ftol, xtol, gtol, rescale, disp, callback, finite_diff_rel_step, maxfun, **unknown_options)
    412             maxfun = max(100, 10*len(x0))
    413 
--> 414     rc, nf, nit, x, funv, jacv = moduleTNC.tnc_minimize(
    415         func_and_grad, x0, low, up, scale,
    416         offset, messages, maxCGit, maxfun,

moduleTNC.pyx in moduleTNC.tnc_minimize()

moduleTNC.pyx in moduleTNC.callback_function()

moduleTNC.pyx in moduleTNC.callback_function()

/var/folders/m8/gwtcncws12jf60xw5n6knwnw0000gn/T/ipykernel_19744/2729457978.py in callback(*a)
      1 from scipy import optimize
      2 def callback(*a):
----> 3   raise RuntimeError("xxx")
      4 
      5 optimize.minimize(lambda x: optimize.rosen(x), [0, 1], method="TNC", callback=callback)

RuntimeError: xxx
```

so it seems like there's no issue with the Exception raising at least. I attached the Python interpreter process to the macOS Instruments program and I can't detect any leaks when the callback exception is raised.